### PR TITLE
librados: implement async watch/unwatch

### PR DESCRIPTION
**Pull request by yuyuyu101**

### CODE REVIEW

**Comment by yuyuyu101** [comment]
done

**Comment by yuyuyu101** [comment]
retest this please

**Comment by dillaman** [comment]
@yuyuyu101 Can you also add an async version of Rados::watch_flush?

**Comment by yuyuyu101** [comment]
@dillaman sure

**Comment by yuyuyu101** [comment]
@dillaman done

File: src/test/librados/watch_notify.cc [hunk 1]
```diff
@@ -593,7 +726,14 @@ TEST_F(LibRadosWatchNotify, WatchNotify2Timeout) {
   ASSERT_GT(rados_watch_check(ioctx, handle), 0);
 
   rados_unwatch2(ioctx, handle);
-  rados_watch_flush(cluster);
+
+  rados_completion_t comp;
+  ASSERT_EQ(0, rados_aio_create_completion(NULL, NULL, NULL, &comp));
+  rados_async_watch_flush(cluster, comp);
```
Comment (dillaman) [inline src/test/librados/watch_notify.cc]: Doesn't build -- should be ```rados_aio_watch_flush```

**Comment by yuyuyu101** [comment]
done

**Comment by dillaman** [comment]
I had this branch merged into wip-jd-testing for a teuthology run: http://pulpito.ceph.com/jdillaman-2016-02-22_09:18:56-rbd-wip-jd-testing---basic-multi/

It appears all clients are locking up on the sync watch_flush call:

> [ubuntu@smithi001 ~]$ ceph osd crush tunables default
> 2016-02-22 17:47:58.631634 7efcce041700  1 librados: starting msgr at :/0
> 2016-02-22 17:47:58.631645 7efcce041700  1 librados: starting objecter
> 2016-02-22 17:47:58.631919 7efcce041700  1 librados: setting wanted keys
> 2016-02-22 17:47:58.631923 7efcce041700  1 librados: calling monclient init
> 2016-02-22 17:47:58.635064 7efcce041700  1 librados: init done
> adjusted tunables profile to default
> 2016-02-22 17:47:58.919154 7efcce041700 10 librados: watch_flush enter
>

**Comment by yuyuyu101** [comment]
@dillaman I agree with your point. But I don't find anywhere we call watch_flush after objecter shutdown

**Comment by dillaman** [comment]
https://github.com/yuyuyu101/ceph/blob/wip-async-watch/src/librados/RadosClient.cc#L334
https://github.com/yuyuyu101/ceph/blob/wip-async-watch/src/librados/RadosClient.cc#L346

**Comment by yuyuyu101** [comment]
oh, sorry. I got it

**Comment by yuyuyu101** [comment]
done

**Comment by yuyuyu101** [comment]
done

**Pull request merged**